### PR TITLE
Cleanup and Bolster System.Runtime tests

### PR DESCRIPTION
**Pull request by hughbe**
Add missing namespaces
Add missing tests and test inputs
Increase code coverage (especially for methods/constructors that throw)
Rename, rewrite and split up several tests

Coming up next (after/if this is merged):
- Rename all the tests to use the standard naming conventions
- Rename the test class files to use the standard naming conventions

### CODE REVIEW

**Comment by hughbe** [comment]
N.B: fixes #738

**Comment by stephentoub** [comment]
Several tests are failing.  These failed in the Windows debug run:
```
12:10:23      System.Runtime.Tests.DoubleTests.TestToString(d: -Infinity, format: "G", provider: NumberFormatInfo { CurrencyDecimalDigits = 2, CurrencyDecimalSeparator = ".", CurrencyGroupSeparator = ",", CurrencyGroupSizes = [3], CurrencyNegativePattern = 0, ... }, expected: "-ì") [FAIL]
12:10:23      System.Runtime.Numerics.Tests  Total: 311, Errors: 0, Failed: 0, Skipped: 0, Time: 13.924s
12:10:23         Assert.Equal() Failure
12:10:23                     (pos 1)
12:10:23         Expected: -ì
12:10:23         Actual:   -Infinity
12:10:23                     (pos 1)
12:10:23         Stack Trace:
12:10:23            d:\j\workspace\dotnet_corefx\windows_nt_debug_prtest\src\System.Runtime\tests\System\Double.cs(367,0): at System.Runtime.Tests.DoubleTests.TestToString(Double d, String format, IFormatProvider provider, String expected)
12:10:23      System.Runtime.Tests.DoubleTests.TestToString(d: Infinity, format: "G", provider: NumberFormatInfo { CurrencyDecimalDigits = 2, CurrencyDecimalSeparator = ".", CurrencyGroupSeparator = ",", CurrencyGroupSizes = [3], CurrencyNegativePattern = 0, ... }, expected: "ì") [FAIL]
12:10:23         Assert.Equal() Failure
12:10:23                    (pos 0)
12:10:23         Expected: ì
12:10:23         Actual:   Infinity
12:10:23                    (pos 0)
12:10:23         Stack Trace:
12:10:23            d:\j\workspace\dotnet_corefx\windows_nt_debug_prtest\src\System.Runtime\tests\System\Double.cs(367,0): at System.Runtime.Tests.DoubleTests.TestToString(Double d, String format, IFormatProvider provider, String expected)
12:10:23      System.Runtime.Tests.SingleTests.TestToString(f: -Infinity, format: "G", provider: NumberFormatInfo { CurrencyDecimalDigits = 2, CurrencyDecimalSeparator = ".", CurrencyGroupSeparator = ",", CurrencyGroupSizes = [3], CurrencyNegativePattern = 0, ... }, expected: "-ì") [FAIL]
12:10:23         Assert.Equal() Failure
12:10:23                     (pos 1)
12:10:23         Expected: -ì
12:10:23         Actual:   -Infinity
12:10:23                     (pos 1)
12:10:23         Stack Trace:
12:10:23            d:\j\workspace\dotnet_corefx\windows_nt_debug_prtest\src\System.Runtime\tests\System\Single.cs(367,0): at System.Runtime.Tests.SingleTests.TestToString(Single f, String format, IFormatProvider provider, String expected)
12:10:23      System.Runtime.Tests.SingleTests.TestToString(f: Infinity, format: "G", provider: NumberFormatInfo { CurrencyDecimalDigits = 2, CurrencyDecimalSeparator = ".", CurrencyGroupSeparator = ",", CurrencyGroupSizes = [3], CurrencyNegativePattern = 0, ... }, expected: "ì") [FAIL]
12:10:23         Assert.Equal() Failure
12:10:23                    (pos 0)
12:10:23         Expected: ì
12:10:23         Actual:   Infinity
12:10:23                    (pos 0)
12:10:23         Stack Trace:
12:10:23            d:\j\workspace\dotnet_corefx\windows_nt_debug_prtest\src\System.Runtime\tests\System\Single.cs(367,0): at System.Runtime.Tests.SingleTests.TestToString(Single f, String format, IFormatProvider provider, String expected)
```
These failed on Ubuntu debug:
```
12:30:53    System.Runtime.Tests.StringTests.TestIndexOfAny_Invalid [FAIL]
12:30:53       Assert.Equal() Failure
12:30:53       Expected: (null)
12:30:53       Actual:   anyOf
12:30:53       Stack Trace:
12:30:54             at System.Runtime.Tests.StringTests.TestIndexOfAny_Invalid()
12:30:54    System.Runtime.Tests.StringTests.TestLastIndexOfAny_Invalid [FAIL]
12:30:54       Assert.Equal() Failure
12:30:54       Expected: (null)
12:30:54       Actual:   anyOf
12:30:54       Stack Trace:
12:30:54             at System.Runtime.Tests.StringTests.TestLastIndexOfAny_Invalid()
12:30:54 Finished:    System.Security.Cryptography.OpenSsl.Tests
12:30:54 
12:30:54 === TEST EXECUTION SUMMARY ===
12:30:54    System.Security.Cryptography.OpenSsl.Tests  Total: 131, Errors: 0, Failed: 0, Skipped: 0, Time: 6.652s
12:30:55    System.Runtime.Tests.StringBuilderTests.TestAppendLine(original: "Hello", value: "abc", expected: "Helloabc\r\n") [FAIL]
12:30:55       Assert.Equal() Failure
12:30:55                         ��� (pos 8)
12:30:55       Expected: Helloabc\r\n
12:30:55       Actual:   Helloabc\n
12:30:55                         ��� (pos 8)
12:30:55       Stack Trace:
12:30:55             at System.Runtime.Tests.StringBuilderTests.TestAppendLine(String original, String value, String expected)
12:30:55    System.Runtime.Tests.StringBuilderTests.TestAppendLine(original: "Hello", value: "", expected: "Hello\r\n") [FAIL]
12:30:55       Assert.Equal() Failure
12:30:55                      ��� (pos 5)
12:30:55       Expected: Hello\r\n
12:30:55       Actual:   Hello\n
12:30:55                      ��� (pos 5)
12:30:55       Stack Trace:
12:30:55             at System.Runtime.Tests.StringBuilderTests.TestAppendLine(String original, String value, String expected)
12:30:55    System.Runtime.Tests.StringBuilderTests.TestAppendLine(original: "Hello", value: null, expected: "Hello\r\n") [FAIL]
12:30:55       Assert.Equal() Failure
12:30:55                      ��� (pos 5)
12:30:55       Expected: Hello\r\n
12:30:55       Actual:   Hello\n
12:30:55                      ��� (pos 5)
12:30:55       Stack Trace:
12:30:55             at System.Runtime.Tests.StringBuilderTests.TestAppendLine(String original, String value, String expected)
12:30:57    System.Runtime.Tests.ArrayTests.TestCreateInstance_Type_IntArray_IntArray_Invalid [FAIL]
12:30:57       Assert.Throws() Failure
12:30:57       Expected: typeof(System.ArgumentOutOfRangeException)
12:30:57       Actual:   typeof(System.OutOfMemoryException): Array dimensions exceeded supported range.
12:30:57       Stack Trace:
12:30:57             at System.Array.InternalCreate(Void* elementType, Int32 rank, Int32* pLengths, Int32* pLowerBounds)
12:30:57             at System.Array.CreateInstance(Type elementType, Int32[] lengths, Int32[] lowerBounds)
12:30:57             at System.Runtime.Tests.ArrayTests.<>c.<TestCreateInstance_Type_IntArray_IntArray_Invalid>b__34_8()
```
I haven't looked at the other legs.

**Comment by hughbe** [comment]
@stephentoub 
the Windows build failures are bizarre - it must be some sort of localization issue! Some of these other test failures suggest that there are inconsistencies between implementations amongst different platforms - nothing major but that should ring some bells!

**Comment by hughbe** [comment]
Just submitted a quick patch to the test failures - I'll check the logs later when all the tests have finished to make sure I didn't muck anything else up ;P

File: src/System.Runtime/tests/Performance/Perf.String.cs [hunk 1]
```diff
@@ -3,102 +3,133 @@
...
         {
             yield return new object[] { 10 };
             yield return new object[] { 100 };
             yield return new object[] { 1000 };
         }
 
-        [Benchmark]
-        [MemberData("TestStringSizes")]
-        public void GetChars(int size)
+        [Benchmark , MemberData("StringSizesTestData")]
```
Comment (eerhardt) [inline src/System.Runtime/tests/Performance/Perf.String.cs]: Is there a coding/style guideline for this?  My personal preference is to have an attribute per line (like the old code), as I think it is easier to read and easier to search for.

File: src/System.Runtime/tests/Performance/Perf.String.cs [hunk 2]
```diff
@@ -123,158 +154,215 @@ public void Format(int numberOfObjects)
...
-        public void op_Equality(int size)
+        [Benchmark, MemberData("StringSizesTestData")]
+        public static void Operator_Equality(int size)
         {
-            PerfUtils utils = new PerfUtils();
-            bool result;
-            string testString1 = utils.CreateString(size);
-            string testString2 = utils.CreateString(size);
+            var utils = new PerfUtils();
+            bool b;
```
Comment (eerhardt) [inline src/System.Runtime/tests/Performance/Perf.String.cs]: I think the old name "result" was better than just "b".

File: src/System.Runtime/tests/Performance/Perf.StringBuilder.cs [hunk 3]
```diff
@@ -2,61 +2,74 @@
...
 {
-    public class Perf_StringBuilder
+    public static class Perf_StringBuilder
     {
         [Benchmark]
-        public void ctor()
+        public static void Ctor()
         {
-            StringBuilder builder;
+            StringBuilder sb;
```
Comment (eerhardt) [inline src/System.Runtime/tests/Performance/Perf.StringBuilder.cs]: Why the naming change? Using abbreviations makes the code harder to read.

File: src/System.Runtime/tests/Performance/Perf.Type.cs [hunk 4]
```diff
@@ -5,36 +5,44 @@
...
 {
-    public class Perf_Type
+    public static class Perf_Type
     {
         [Benchmark]
-        public void GetTypeFromHandle()
+        public static void GetTypeFromHandle()
         {
-            RuntimeTypeHandle type1 = typeof(int).TypeHandle;
+            RuntimeTypeHandle th = typeof(int).TypeHandle;
```
Comment (eerhardt) [inline src/System.Runtime/tests/Performance/Perf.Type.cs]: Same question here on the naming change - 'type' or 'typeHandle' are a lot better names than 'th' IMO.

File: src/System.Runtime/tests/Performance/Perf.String.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Runtime/tests/Performance/Perf.String.cs]: I just follow what the majority of tests in the project follow - in this case its the new one liner

File: src/System.Runtime/tests/Performance/Perf.StringBuilder.cs (hunk 3 above)
Comment (hughbe) [inline src/System.Runtime/tests/Performance/Perf.StringBuilder.cs]: I just follow what the majority of tests in the project follow - basically all the StringBuilder.cs test files use sb so I was being consistent

**Comment by hughbe** [comment]
Okay everythings working! 
Also I have a question: is there a way to check out code coverage for System.Runtime tests? That would be really helpful to add missing tests!!

**Comment by stephentoub** [comment]
>  is there a way to check out code coverage for System.Runtime tests?

Yes, but it's much more involved.  Basically:
1. Do a normal code coverage run for System.Runtime.  This will populate a folder ```corefx\bin\tests\Windows_NT.AnyCPU.Debug\System.Runtime.Tests\dnxcore50``` will lots of stuff and also spit out to the console the commands that were used to invoke OpenCover and ReportGenerator.  The actual code coverage report generated doesn't matter.
2. Build coreclr locally in debug.
3. Copy the debug binaries (e.g. mscorlib.dll, coreclr.dll, mscorlib.pdb, etc.) from the coreclr build to that dnxcore50 folder, overwriting files already there.
4. Delete the mscorlib.ni.dll file from that dnxcore50 folder.
5. In that dnxcore50 folder, run the OpenCover command, then run the ReportGenerator command.

**Comment by stephentoub** [comment]
@hughbe, can you rebase this? Thanks.

**Comment by stephentoub** [comment]
@hughbe, GitHub is still reporting "This branch has conflicts that must be resolved".

**Comment by hughbe** [comment]
yeah I'm just working on rebasing it but accidently force pushed stuff! I'll ping you when I'm done :)

**Comment by hughbe** [comment]
Whoops

**Comment by hughbe** [comment]
Okay I accidently closed this

**Comment by hughbe** [comment]
@stephentoub rebased!

**Comment by hughbe** [comment]
@stephentoub done! for some reason some files (e.g. String.cs) cannot even be viewed on their own (it is the only file in the commit for it)! I'm not sure what to do there

**Comment by hughbe** [comment]
@stephentoub 
I fixed the latest merge conflict. Sorry to be pushy but it is effort rebasing and recommitting each time a merge conflict occurs, so I'd appreciate it if you reviewed this soon :p

Also, you didnt answer my question requesting other areas of the test project to work on in a future PR! Thanks

**Comment by stephentoub** [comment]
@hughbe, thanks for your patience.  I'll try to find time to review tomorrow.

**Comment by stephentoub** [comment]
I sat down to start looking at this, and began by running the tests, but I get a failure:
```
System.Runtime.Tests.DateTimeOffsetTests.TestParseExact_String_String_FormatProvider [FAIL]
        System.FormatException : The UTC representation of the date falls outside the year range 1-9999.
        Stack Trace:
              at System.DateTimeParse.ParseExact(String s, String format, DateTimeFormatInfo dtfi, DateTimeStyles style, TimeSpan& offset)
              at System.DateTimeOffset.ParseExact(String input, String format, IFormatProvider formatProvider, DateTimeStyles styles)
              at System.DateTimeOffset.ParseExact(String input, String format, IFormatProvider formatProvider)
           c:\Users\stoub\Source\repos\corefx\src\System.Runtime\tests\System\DateTimeOffset.cs(549,0): at System.Runtime.Tests.DateTimeOffsetTests.TestParseEx
  act_String_String_FormatProvider()
```

**Comment by stephentoub** [comment]
General: There are a bunch of places where you've replaced code like ```DateTime dt;``` with ```DateTime dateTime```... why?  It's pretty common to name variables in this way, especially when it's very simple code like in these benchmarks:
```C#
        [Benchmark]
        public void GetNow()
        {
            DateTime dt;
            foreach (var iteration in Benchmark.Iterations)
                using (iteration.StartMeasurement())
                    for (int i = 0; i < 10000; i++)
                    {
                        dt = DateTime.Now; dt = DateTime.Now; dt = DateTime.Now;
                        dt = DateTime.Now; dt = DateTime.Now; dt = DateTime.Now;
                        dt = DateTime.Now; dt = DateTime.Now; dt = DateTime.Now;
                    }
        }
```
I'm not clear on why it's better as ```dateTime``` instead of ```dt```... seems like a change just for change's sake.  I can fully imagine someone else coming along in the future and changing it back :wink:

**Comment by stephentoub** [comment]
You made some changes like changing:
```C#
[Benchmark]
[MemberData("StringSizesTestData")]
```
to:
```C#
[Benchmark , MemberData("StringSizesTestData")]
```
I'd prefer if in general we kept these separate on multiple lines.  I understand there are some places in the code base where we don't, but if we're going to standardize on one approach, I much prefer it be each attribute by itself (that doesn't mean we should go through the code and update all occurrences where it's not, though).

**Comment by stephentoub** [comment]
You changed a bunch of test methods to be static.  I'm still not clear on the value of doing that.  Are you noticing substantial perf throughput improvements on the tests as a result?  If not, let's revert all of that.  There is benefit to having them be instance, in that it makes it easier to have test classes inherit from other test classes, for a test class to have setup/teardown support, etc., and since we need instance test methods in many places as a result, I'd prefer striving for consistency of having test methods be instance rather than some be instance and some be static, and having to change static back to instance when we want to take advantage of such services.  If new test methods get written as as statics, that's one thing, but to go through and programmatically convert a bunch of instance test methods to static is another.

**Comment by stephentoub** [comment]
Nit: DirectoryNotFoundException.cs L14. "excpetion" => "exception"

**Comment by stephentoub** [comment]
DirectoryNotFoundException.Interop.cs L16.  Seems like this would be better as a [Theory].

**Comment by stephentoub** [comment]
FileLoadException.Interop.cs L16.  Same, seems like this would be better as a [Theory].

**Comment by stephentoub** [comment]
StringBuilder.cs L700.  I'm confused by this test... in a bunch of places, you append to the StringBuilder but then immediately overwrite the StringBuilder and don't first verify anything about its contents.  This is made more difficult to see because you reuse the same StringBuilder variable in multiple places.

**Comment by stephentoub** [comment]
StringBuilder.cs 844.  Same thing... you create a StringBuilder, append to it, and then immediately overwrite it.  Did you mean to add an assert here?

**Comment by stephentoub** [comment]
Activator.cs L116.  Changing this type from public to private changes all of the tests that use it, since visibility plays a key role in reflection.

**Comment by stephentoub** [comment]
Array.cs L395.  What is this test testing?  I'm unclear how it relates to the test's name ("TestCopy_GCReferenceArray").  Is it just meant to be testing copying an array with reference type instances?  If so, ok, I just find the name confusing.

**Comment by stephentoub** [comment]
Array.cs L428.  This test seems like it's really several distinct tests... I'd suggest splitting it up into different Fact methods.

**Comment by stephentoub** [comment]
Array.cs L790 (TestEmpty).  The test comparing two calls to Empty should be using Assert.Same rather than Assert.Equal.  We not only want them to be equal in contents, we want them to be the same exact object.

**Comment by stephentoub** [comment]
Array.cs L1292.  I'm finding this test confusing... could you split this into multiple tests (even if some bail early because the conditions for it aren't met, though you could avoid that by splitting up the test data as well)?  Also, I'm not clear on why sortedStringArray is being assigned a new empty array... it looks like every usage of it first overwrites it.

**Comment by stephentoub** [comment]
Array.cs L1477.  Same comment about this test (TestSort_Array_Array) being quite sprawling and confusing.  Please split it up.  This is another case where reusing the same variables over and over means a reader needs to follow when those variables are assigned and whether a previous aspect of the test is important for a later aspect of the test.

**Comment by stephentoub** [comment]
(I reviewed up through Buffer.cs... I'll come back later to keep reviewing more.)

**Comment by hughbe** [comment]
Awesome thanks for this review. I'll follow up on these tomorrow. 

R.e. The dateTime -> dt change, I simply followed what I consider better coding guidelines - and also I was told to correct instances where I used sb instead of stringBuilder so I assume this rationale applied to all types. 

R.e the benchmark stuff, I wanted to be consistent with the layout of all this. I know see the pros in the two line layout. Won't happen again in the future!

R.e. The static changes. All other methods were static. As a person who values consistency, I changed instance methods to the majority (static)

Thanks so much, sorry if there's a lot of stuff I need to fix! :)

**Comment by hughbe** [comment]
I just pushed a PR feedback commit. Hopefully this will fix your test failure (that isn't occuring on my machine), and make some of the tests you had a hard time understanding  more clear (I could barely understand them looking back now)

**Comment by hughbe** [comment]
Oh dear, another merge conflict :( 

I'm really sorry if this means you have to re-review things after I fix the merge conflcit

**Comment by hughbe** [comment]
Actually something just occurred to me @stephentoub.

Would you like me to split up this PR into multiple smaller PRs so that it's easier for you to review and easier for me to fix rebas issues ;)

**Comment by stephentoub** [comment]
> Would you like me to split up this PR into multiple smaller PRs so that it's easier for you to review and easier for me to fix rebas issues 

That would be much better. Thanks!

**Comment by hughbe** [comment]
Yeah that works pretty well. 

I'll do that tomorrow afternoon.

**Comment by hughbe** [comment]
Closing this, as this is being split up into smaller PRs.

**Pull request closed without merging**